### PR TITLE
Extend OWM-Plugin to use OWM onecall API, add irrigation features, improve Webinterface

### PR DESCRIPTION
**Pull request by jentz1986**
Refer https://knx-user-forum.de/forum/supportforen/smarthome-py/1246998-support-thread-zum-openweathermap-plugin?p=1671786#post1671786 for discussion

### CODE REVIEW

**Comment by jentz1986** [comment]
Depending on release dates and feedback in Forum: I think this version can be merged into develop.

File: openweathermap/user_docu.rst [hunk 1]
```diff
@@ -0,0 +1,1051 @@
...
+
+
+Funktionaliät
+=============
+
+Allgemeine Information
+----------------------
+
+Die von OpenWeatherMap bereitgestellten Daten werden mit Hilfe eines "Matchstrings" zu den Items verknüpft. Der "Matchstring" definiert den Pfad zur Datenquelle innerhalb des Datenfeldes.
+Die Liste der unterstützten APIs stellt die Namen der "data-source-keys" dar, die im Plugin verwendet werden. Das Ergebnis des letzten Datenabrufes wird im WebIF des Pluigns dargestellt.
```
Comment (bmxp) [inline openweathermap/user_docu.rst:32]: Plugins

File: openweathermap/user_docu.rst [hunk 2]
```diff
@@ -0,0 +1,1051 @@
+==============
```
Comment (bmxp) [inline openweathermap/user_docu.rst:1]: user_doc.rst instead of user_doc**u**.rst please :-)

File: openweathermap/user_docu.rst [hunk 3]
```diff
@@ -0,0 +1,1051 @@
...
+
+Funktionaliät
+=============
+
+Allgemeine Information
+----------------------
+
+Die von OpenWeatherMap bereitgestellten Daten werden mit Hilfe eines "Matchstrings" zu den Items verknüpft. Der "Matchstring" definiert den Pfad zur Datenquelle innerhalb des Datenfeldes.
+Die Liste der unterstützten APIs stellt die Namen der "data-source-keys" dar, die im Plugin verwendet werden. Das Ergebnis des letzten Datenabrufes wird im WebIF des Pluigns dargestellt.
+Das Plugin lädt, abhängig von dem in der Itemkonfiguration definierten "Matchstrings", die notwendigen Daten. Ist kein entsprechender "Matching" definiert, wird die jeweilige API nicht abgerufen.
```
Comment (bmxp) [inline openweathermap/user_docu.rst:33]: Matching = Matchstring?

File: openweathermap/user_docu.rst [hunk 4]
```diff
@@ -0,0 +1,1051 @@
...
+
+.. code:: html
+
+    {{ status.activelist('', 'weather.alerts', 'event', 'start', 'description', '') }}
+
+
+Matchstring Fehlerbehandlung
+----------------------------
+Das typische Prüfen der Matchstrings wird bei die Wurzel der JSON-Antwort des API-Abrufes beginnen und dann dem im Matchstring definierten "Pfades" folgend die entsprechenden Daten aus dem JSON dem Item zuweisen.
+Wenn der nächste Knoten entlang dieses "Pfades" nicht erreicht werden kan, wird ein ERROR geloggt. Typischerweise entsteht das durch Schreibfehler oder fehlender/falsche Indizees bei Listen.
```
Comment (bmxp) [inline openweathermap/user_docu.rst:287]: Indizes statt Indize**e**s

**Review by bmxp — CHANGES_REQUESTED** [review]
Ich habe ein paar Tippfehler beim schnellen überfliegen gefunden und ein paar andere Kleinigkeiten. Danke erstmal für die Erweiterung und die Doku

File: openweathermap/user_docu.rst [hunk 5]
```diff
@@ -0,0 +1,1051 @@
...
+Nachfolgend noch einige Zusatzinformationen.
+
+
+Items
+-----
+
+Für die Nutzung des Plugins muss in den entsprechenden Items das Attribute ``owm_matchstring`` konfiguriert werden. Optional kann dass Attribut ``owm_match_prefix`` verwendet werden.
+Dieser String wird dem ``owm_matchstring`` vorangestellt und erlaubt eine bessere Struktur bzw. eine einfachere Definition von structs.
+
+.. note:: Die korrekte Definition und Verwendung der Instanz des Plugins ist für die einwandfreie Funktion des Plugins notwendig. In den Beispielen lautet der Name der Instanz **home**
```
[the remark this answers is not in the document]
Comment (jentz1986) [reply openweathermap/user_docu.rst]: Vielen Dank für den Hinweis, habe ich dann so mal in die Doku übernommen. Der Betrieb geht natürlich auch ohne Instanzdefinition, also nur mit der Defaultinstanz

File: openweathermap/user_docu.rst [hunk 6]
```diff
@@ -0,0 +1,1051 @@
...
+        owm_match_prefix@home: day/1
+        owm_matchstring@home: /weather/0/description
+
+        temp_night:
+            type: num
+            remark: here the match-string is compiled as day/1/temp/night, inheriting the prefix from the parent-element.
+            owm_match_prefix@home: ../.
+            owm_matchstring@home: /temp/night
+
+Hier nachfolgend eine komplette item.yaml für die Anwendung des Plugins. Die Instanz (hier **home**) als auch der Plugin-Name (hier **_priv_openweathermap**) ist gemäß Eurer Definition anzupassen.
```
[the remark this answers is not in the document]
Comment (jentz1986) [reply openweathermap/user_docu.rst]: Ja.

**Comment by psilo909** [comment]
When testing with my initial items, two are not working anymore. Acc. to OWM docu, this is because the onecall api does not support them anymore. perhaps its worth documenting that? 

             `   grnd_level:
                    type: num
                    owm_matchstring@home: main/grnd_level

                sea_level:
                    type: num
                    owm_matchstring@home: main/sea_level`

I am now reconfiguring the plugin and testing it for some time. more later

**Comment by psilo909** [comment]
Got one exception in my logs recently:

2021-11-03  01:01:16 ERROR    plugins.openweathermap_home.plugins.openweathermap Method plugins.openweathermap_home.plugins.openweathermap exception: Unterminated string starting at: line 1 column 3525 (char 3524)
> Traceback (most recent call last):
>   File "/python/smarthome/lib/scheduler.py", line 670, in _task
>     obj()
>   File "/python/smarthome/plugins/openweathermap/__init__.py", line 214, in _update_loop
>     self._update()
>   File "/python/smarthome/plugins/openweathermap/__init__.py", line 392, in _update
>     self._download_data()
>   File "/python/smarthome/plugins/openweathermap/__init__.py", line 223, in _download_data
>     only_if=self._request_forecast)
>   File "/python/smarthome/plugins/openweathermap/__init__.py", line 699, in __query_api_if
>     self.__query_api(data_source_key, delta_t)
>   File "/python/smarthome/plugins/openweathermap/__init__.py", line 720, in __query_api
>     json_obj = response.json()
>   File "/usr/local/lib/python3.7/dist-packages/requests/models.py", line 910, in json
>     return complexjson.loads(self.text, **kwargs)
>   File "/usr/lib/python3.7/json/__init__.py", line 348, in loads
>     return _default_decoder.decode(s)
>   File "/usr/lib/python3.7/json/decoder.py", line 337, in decode
>     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
>   File "/usr/lib/python3.7/json/decoder.py", line 353, in raw_decode
>     obj, end = self.scan_once(s, idx)
> json.decoder.JSONDecodeError: Unterminated string starting at: line 1 column 3525 (char 3524)

**Comment by psilo909** [comment]
@jentz1986 anything new here concerning my comments above? I can go on with a review (and i am still using the plugin in production) - but i think its more usefull the take on the things a found so far.

**Comment by jentz1986** [comment]
> When testing with my initial items, two are not working anymore. Acc. to OWM docu, this is because the onecall api does not support them anymore. perhaps its worth documenting that?
> 
> ```
>          `   grnd_level:
>                 type: num
>                 owm_matchstring@home: main/grnd_level
> 
>             sea_level:
>                 type: num
>                 owm_matchstring@home: main/sea_level`
> ```

These fields are in fact still there, but they are not delivered always. Not sure how to work with that. I think the more general current/pressure would do the job. So I do not see any chance to deal with that.

> Also the configured rain layer thing is not working for me anymore (no issue, was not using this anyway):

You have been using the parameter names x, y and z. These are now either auto-configured from the SHNG geo-coordinates configuration, or you supply them all as owm_coord_x, owm_coord_y, owm_coord_z


> Got one exception in my logs recently:
> 
> 2021-11-03 01:01:16 ERROR plugins.openweathermap_home.plugins.openweathermap Method plugins.openweathermap_home.plugins.openweathermap exception: Unterminated string starting at: line 1 column 3525 (char 3524)
If had this some times in my log - seems to be an issue with incomplete JSONs from the source. Ideas?

**Comment by psilo909** [comment]
@jentz1986 thx! i check the latest changes the next days!

**Comment by psilo909** [comment]
Do we need the openweathermap/.gitignore  file in the PR? normally we dont have any in our plugin folders

**Comment by psilo909** [comment]
> > When testing with my initial items, two are not working anymore. Acc. to OWM docu, this is because the onecall api does not support them anymore. perhaps its worth documenting that?
> > ```
> >          `   grnd_level:
> >                 type: num
> >                 owm_matchstring@home: main/grnd_level
> > 
> >             sea_level:
> >                 type: num
> >                 owm_matchstring@home: main/sea_level`
> > ```
> 
> These fields are in fact still there, but they are not delivered always. Not sure how to work with that. I think the more general current/pressure would do the job. So I do not see any chance to deal with that.
> 

perhaps you can just add a comment to the two fields in the documentation and do not add them to any of your structs

**Comment by psilo909** [comment]
> If had this some times in my log - seems to be an issue with incomplete JSONs from the source. Ideas?

i will observe it.. perhaps a try catch?

**Comment by psilo909** [comment]
> You have been using the parameter names x, y and z. These are now either auto-configured from the SHNG geo-coordinates configuration, or you supply them all as owm_coord_x, owm_coord_y, owm_coord_z

removed the coordinates from the items and its fine now. perhaps its worth mentioning (in the documentation or the support thread) once we merge the plugins, that a migration here is needed (ok i see the attributes in the new item_attributes, but i think i am not the only one missing this

**Comment by psilo909** [comment]
Don't directly see other issues. Many of my minor comments above are still open. I will have a final look once the next PR is in. I think its only a matter of 15 minutes to address those. In case of questions, pls comment here.

**Comment by psilo909** [comment]
@bmxp  can you check if your issues have been addressed?

**Review by bmxp — APPROVED** [review]

**Comment by bmxp** [comment]
I just overflew the changes and found nothing that needs to be changed urgently.
If I could add a wish for the future then I would like to have a crontab implemented that is more precise to the pull times instead of a cycle which is a somewhat moving target. Background of this is the need of forecast values for wind, temperature and wind speed  for heating pre calculation.

**Comment by jentz1986** [comment]
> If I could add a wish for the future then I would like to have a crontab implemented that is more precise to the pull times instead of a cycle which is a somewhat moving target.

Feel free to raise an issue in the other repo: https://github.com/jentz1986/shng-owm-plugin/issues/new - that is the main reason for the other repo, aside from the easy shareability of in-between releases with others.

**Comment by jentz1986** [comment]
> Do we need the openweathermap/.gitignore file in the PR? normally we dont have any in our plugin folders

I needed it to not upload pycache from my environment. I can remove it again if this is disturbing any workflow.

**Comment by jentz1986** [comment]
> > > When testing with my initial items, two are not working anymore. Acc. to OWM docu, this is because the onecall api does not support them anymore. perhaps its worth documenting that?
> > > ```
> > >          `   grnd_level:
> > >                 type: num
> > >                 owm_matchstring@home: main/grnd_level
> > > 
> > >             sea_level:
> > >                 type: num
> > >                 owm_matchstring@home: main/sea_level`
> > > ```
> > 
> > 
> > These fields are in fact still there, but they are not delivered always. Not sure how to work with that. I think the more general current/pressure would do the job. So I do not see any chance to deal with that.
> 
> perhaps you can just add a comment to the two fields in the documentation and do not add them to any of your structs

Hmm, I scanned plugin.yaml and all rst-files. I could not find any hint on these two fields anywhere. I can add a section like "things that worked in the past, but now not anymore" to the documentation - maybe as well with some deprecation hints.

**Comment by jentz1986** [comment]
> > If had this some times in my log - seems to be an issue with incomplete JSONs from the source. Ideas?
> 
> i will observe it.. perhaps a try catch?

I considered that - but with that one would lose the original error-message. I considered this to be a valuable information - even from inexperienced users. Do we have some rule for when to capture (and hide from the user) what?

**Comment by jentz1986** [comment]
> > You have been using the parameter names x, y and z. These are now either auto-configured from the SHNG geo-coordinates configuration, or you supply them all as owm_coord_x, owm_coord_y, owm_coord_z
> 
> removed the coordinates from the items and its fine now. perhaps its worth mentioning (in the documentation or the support thread) once we merge the plugins, that a migration here is needed (ok i see the attributes in the new item_attributes, but i think i am not the only one missing this

Tried to create a migration guide. Did I miss something there?

**Comment by psilo909** [comment]
final check of the PR starting now. I try to give feedback until today evening or latest tomorrow.

**Comment by psilo909** [comment]
> Tried to create a migration guide. Did I miss something there?

looks like i missed checking the documentation. all fine ;)

**Comment by psilo909** [comment]
Hi, these 3 minor issues are still open for me - i should have pulled latest version of the PR for the check:
image

**Comment by psilo909** [comment]
one more thing: i just realized you are using datatables and are correctly referencing the version delivered with http module (http://.../gstatic/datatables/datatables.min.js)

Can you remove the redundantly shipped one in the plugin itself?

**Comment by msinn** [comment]
Are still larger things to be done or can the PR be merged to be included in the upcoming release?

**Comment by psilo909** [comment]
I can fix the typos. The largest thing seems to be a structure that does not work. For me it can be merged and corrected by a new pr

**Comment by jentz1986** [comment]
> one more thing: i just realized you are using datatables and are correctly referencing the version delivered with http module (http://.../gstatic/datatables/datatables.min.js)
> 
> Can you remove the redundantly shipped one in the plugin itself?

Sure I can (in the future, as mentioned in the files)... but as of now this is not present in the master-branch / Release 1.8.2 therefor I had implemented a "switch" functionality to draw datatables from the standard location if it is present. if not the plugin uses its own sauce. Otherwise this plugin would not be compatible with the currently released version, and the sh min version in plugin.yaml would have to carry a different value...

**Comment by jentz1986** [comment]
> Hi, these 3 minor issues are still open for me - i should have pulled latest version of the PR for the check: image

Indeed, there are two things:

- Thank you for pointing at lcoals again - fixed it. (I guess no one would ever use that)
- the readme.rst is actually a working draft version of the collaboration with @sisamiwe . We had replaced it with a user_doc.rst in English (his work, not mine!!!) and readme_en.rst to have a readme still present in the repo. I have removed the file now completely to avoid future misunderstandings.

**Comment by jentz1986** [comment]
Tried to fix ALL remarks by @psilo909 at this point. Thanks for all the feedback!

**Comment by msinn** [comment]
readme.rst is a non supported file for documentation
Documentation that is to be included in the smarthomeng user documentation uas to reside in user_doc.rst and MUST be written in german.

**Comment by msinn** [comment]
I added some fixes/extensions to the plugin and bumped the version to 1.8.3

**Pull request merged**